src/groups/mqb/mqbblp/mqbblp_rootqueueengine.cpp

```diff
@@ -1619,6 +1619,132 @@ void RootQueueEngine::onTimer(bsls::Types::Int64 currentTimer)
     d_consumptionMonitor.onTimer(currentTimer);
 }
 
+bsl::ostream&
+RootQueueEngine::logAppSubscriptionInfo(bsl::ostream&           stream,
+                                        const mqbu::StorageKey& appKey) const
+{
+    // executed by the *QUEUE DISPATCHER* thread
+
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_queueState_p->queue()->dispatcher()->inDispatcherThread(
+        d_queueState_p->queue()));
+
+    // Get AppState by appKey.
+    Apps::const_iterator cItApp = d_apps.findByKey2(AppKeyCount(appKey, 0));
+    if (cItApp == d_apps.end()) {
+        BALL_LOG_WARN << "No app found for appKey: " << appKey;
+        stream << "\nSubscription info: no app found for appKey: " << appKey;
+        return stream;  // RETURN
+    }
+
+    const AppStateSp& app = cItApp->value();
+    return logAppSubscriptionInfo(stream, app);
+}
+
+bsl::ostream&
+RootQueueEngine::logAppSubscriptionInfo(bsl::ostream&     stream,
+                                        const AppStateSp& appState) const
+{
+    mqbi::Storage* const storage = d_queueState_p->storage();
+
+    // Log un-delivered messages info
+    stream << "\nFor appId: " << appState->appId() << "\n\n";
+    stream << "Put aside list size: "
+           << bmqu::PrintUtil::prettyNumber(static_cast<bsls::Types::Int64>(
+                  appState->putAsideListSize()))
+           << '\n';
+    stream << "Redelivery list size: "
+           << bmqu::PrintUtil::prettyNumber(static_cast<bsls::Types::Int64>(
+                  appState->redeliveryListSize()))
+           << '\n';
+    stream << "Number of messages: "
+           << bmqu::PrintUtil::prettyNumber(
+                  storage->numMessages(appState->appKey()))
+           << '\n';
+    stream << "Number of bytes: "
+           << bmqu::PrintUtil::prettyBytes(
+                  storage->numBytes(appState->appKey()))
+           << "\n\n";
+
+    // Log consumer subscriptions
+    mqbblp::Routers::QueueRoutingContext& routingContext =
+        appState->routing()->d_queue;
+    mqbcmd::Routing routing;
+    routingContext.loadInternals(&routing);
+    const bsl::vector<mqbcmd::SubscriptionGroup>& subscrGroups =
+        routing.subscriptionGroups();
+    if (!subscrGroups.empty()) {
+        // Limit to log only k_EXPR_NUM_LIMIT expressions
+        static const size_t k_EXPR_NUM_LIMIT = 50;
+        bmqu::MemOutStream  ss(d_allocator_p);
```

Comment:
Why not output to the `stream` directly and avoid using `ss`

Reply:
In line 1692 there is checking of expressions number and there is different caption depending on check result to explain user that limited number of expressions is printed, before printing list of expressions. Is there better way to get number of non-empty subscriptions?

Reply:
Can we not use `subscrGroups.size()` upfront and log the header line before iterating?

Reply:
I noticed that subscrGroups container is not empty even if there is no subscriptions, but it contains empty expressions (that's why I use `if (!cIt->expression().empty())` and `exprNum` counter to get real number of expressions.

Reply:
@dorjesinpo got rid of using interim `ss`, added printing of empty subscription. Tested that there is no duplication in `subscrGroups`.